Condvar: implement wait_timeout for targets without threads

surban opened the pull request:
This always falls back to sleeping since there is no way to notify a condvar on a target without threads.

Even on a target that has no threads the following code is a legitimate use case:

```rust
use std::sync::{Condvar, Mutex};
use std::time::Duration;

fn main() {
    let cv = Condvar::new();
    let mutex = Mutex::new(());
    let mut guard = mutex.lock().unwrap();

    cv.notify_one();

    let res;
    (guard, res) = cv.wait_timeout(guard, Duration::from_secs(3)).unwrap();
    assert!(res.timed_out());
}
```

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r\? <reviewer name> (with the `\` removed)
-->
<!-- homu-ignore:end -->

Darksonn commented on library/std/src/sys/sync/condvar/no_threads.rs:
```diff
@@ -1,25 +1,50 @@
-        panic!("condvar wait not supported")
+        let mut value = self.value.borrow_mut();
+        if *value == 0 {
+            // Since the target does not have threads,
+            // no notification can ever arrive.
+            loop {}
```
Most wasm primitives panic when they hit a situation that would otherwise sleep forever.

surban replied:
I changed it to a panic.

tgross35 commented on library/std/src/sys/sync/condvar/no_threads.rs:
```diff
@@ -1,25 +1,48 @@
 
-pub struct Condvar {}
+pub struct Condvar {
+    value: RefCell<u32>,
```
It looks like the value is always accessed via `borrow_mut`, can't this be a `Cell` rather than `RefCell`?

tgross35 replied:
Also could this get a descriptive name like `unused_notifications`?

tgross35 commented on library/std/src/sys/sync/condvar/no_threads.rs:
```diff
@@ -1,25 +1,48 @@
     #[inline]
-    pub fn notify_one(&self) {}
+    pub fn notify_one(&self) {
+        *self.value.borrow_mut() += 1;
+    }
 
     #[inline]
-    pub fn notify_all(&self) {}
+    pub fn notify_all(&self) {
+        *self.value.borrow_mut() += 1;
+    }
```
Maybe the type should be `Saturating<u32>` so this has no chance of wrapping to 0.

tgross35 commented:
This seems reasonable to me but since this is new user-visible behavior, requesting libs-api to take a look.

@rustbot label +I-libs-api-nominated

surban committed:
Implement Condvar::wait_timeout for targets without threads

This always falls back to sleeping since there is no way
to notify a condvar on a target without threads.

surban commented:
@tgross35 

There was a misunderstanding about buffering in Condvars. Indeed [the documentation](https://doc.rust-lang.org/std/sync/struct.Condvar.html) states:

> Calls to `notify_one` are not buffered in any way.
> Calls to `notify_all()` are not buffered in any way.

Thus my previous approach using a counter makes no sense and indeed `Condvar::wait` should always panic since there is no way to receive a notification while waiting.

However, `Condvar::wait_timeout` should always wait for the specified timeout and then return with a timed out state.

I have modified the PR and description accordingly.

m-ou-se commented:
@bors r+

bors commented:
:pushpin: Commit 45c7ddfea6ba9980a1c6f4f9bd9c821e56af035e has been approved by `m-ou-se`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=m-ou-se 45c7ddfea6ba9980a1c6f4f9bd9c821e56af035e -->
<!-- homu: {"type":"Approved","sha":"45c7ddfea6ba9980a1c6f4f9bd9c821e56af035e","approver":"m-ou-se","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
